metaflow/plugins/metadata/local.py

```diff
@@ -115,6 +119,88 @@ def register_metadata(self, run_id, step_name, task_id, metadata):
         }
         self._save_meta(meta_dir, metadict)
 
+    @classmethod
+    def _mutate_user_tags_for_run(
+        cls, flow_id, run_id, tags_to_add=None, tags_to_remove=None
+    ):
+        MutationResult = namedtuple(
+            "MutationResult", field_names="tags_are_consistent tags"
+        )
+
+        def _optimistically_mutate():
+            # get existing tags
+            run = LocalMetadataProvider.get_object(
+                "run", "self", {}, None, flow_id, run_id
+            )
+            if not run:
+                raise MetaflowTaggingError(
+                    msg="Run not found (%s, %s)" % (flow_id, run_id)
+                )
+            existing_user_tag_set = frozenset(run["tags"])
+            existing_system_tag_set = frozenset(run["system_tags"])
+            tags_to_remove_set = frozenset(tags_to_remove)
+            # make sure no existing system tags get added as a user tag
+            tags_to_add_set = frozenset(tags_to_add) - existing_system_tag_set
+
+            # from this point on we work with sets of tags only
+
+            if tags_to_remove_set & existing_system_tag_set:
+                raise MetaflowTaggingError(
+                    msg="Cannot remove a tag that is an existing system tag (%s)"
+                    % str(sorted(tags_to_remove_set & existing_system_tag_set))
+                )
+
+            # remove tags first, then add
+            next_user_tags_set = (
+                existing_user_tag_set - tags_to_remove_set
+            ) | tags_to_add_set
+            # we think it will be a no-op, so let's return right away
+            if next_user_tags_set == existing_user_tag_set:
+                return MutationResult(
+                    tags=next_user_tags_set,
+                    tags_are_consistent=True,
+                )
+
+            validate_tags(next_user_tags_set, existing_tags=existing_user_tag_set)
```

Comment:
`LocalMetadataProvider` validates tag set sizing.  Yes we will implicitly check all the tag values again.  Not a big deal, keeps code simple.